File: src/services/AutoUpdater.ts
import { errorLogger } from "../utils/error-logger.ts";
import { logger } from "../utils/logger.ts";

// Web-compatible interfaces for PWA/desktop app updates
export interface AppLike {
  getVersion(): string;
}

export interface UpdateNotificationCallback {
  (channel: string, data?: any): void;
}

export interface AutoUpdaterConfig {
  checkOnStartup?: boolean;
  githubRepo?: string; // e.g., "owner/repo"
  updateCheckInterval?: number; // in milliseconds
  channel?: "stable" | "beta" | "alpha";
}

export interface UpdateStatus {
  currentVersion: string;
  latestVersion?: string;
  checking: boolean;
  updateAvailable: boolean;
  downloading: boolean;
  updateDownloaded: boolean;
  downloadProgress?: number;
  error?: string;
}

export interface GitHubAsset {
  id: number;
  name: string;
  browser_download_url: string;
  size: number;
}

export interface GitHubRelease {
  id: number;
  tag_name: string;
  name: string;
  body: string;
  published_at: string;
  html_url: string;
  prerelease: boolean;
  assets: GitHubAsset[];
}

export interface UpdateCheckResult {
  updateAvailable: boolean;
  currentVersion: string;
  latestVersion?: string;
  releaseNotes?: string;
  downloadUrl?: string;
  error?: string;
}

export interface UpdateDownloadResult {
  success: boolean;
  error?: string;
}

export class AutoUpdater {
  private config: AutoUpdaterConfig;
  private notificationCallback: UpdateNotificationCallback | null = null;
  private status: UpdateStatus;
  private updateCheckTimer: NodeJS.Timeout | null = null;
  private downloadProgressCallbacks: Array<(percent: number) => void> = [];

  constructor(app: AppLike, config: AutoUpdaterConfig = {}) {
    this.config = {
      checkOnStartup: true,
      channel: "stable",
      ...config,
    };

    this.status = {
      currentVersion: app.getVersion(),
      checking: false,
      updateAvailable: false,
      downloading: false,
      updateDownloaded: false,
    };

    this.configure();
    this.setupPeriodicChecks();
  }

  /**
   * Configure the updater for web-based checking
   */
  private configure(): void {
    const source = this.config.githubRepo
      ? `GitHub releases for ${this.config.githubRepo}`
      : "GitHub releases (no repo configured)";
    logger.info("[AutoUpdater] Configured for", source);
  }

  /**
   * Set up periodic update checks
   */
  private setupPeriodicChecks(): void {
    if (
      this.config.updateCheckInterval &&
      this.config.updateCheckInterval > 0
    ) {
      this.updateCheckTimer = setInterval(() => {
        this.checkForUpdates().catch((error) => {
          logger.error("[AutoUpdater] Periodic check failed:", error);
        });
      }, this.config.updateCheckInterval);
    }
  }

  /**
   * Notify via callback of update events
   */
  private notifyCallback(channel: string, data?: any): void {
    if (this.notificationCallback) {
      this.notificationCallback(channel, data);
    }
  }

  /**
   * Fetch latest release from GitHub
   */
  private async fetchLatestRelease(): Promise<GitHubRelease | null> {
    if (!this.config.githubRepo) {
      throw new Error("GitHub repository not configured");
    }

    try {
      const response = await fetch(
        `https://api.github.com/repos/${this.config.githubRepo}/releases/latest`,
      );

      if (!response.ok) {
        throw new Error(`GitHub API error: ${response.status}`);
      }

      const release: GitHubRelease = await response.json();
      return release;
    } catch (error) {
      logger.error("[AutoUpdater] Failed to fetch release:", error);
      throw error;
    }
  }

  /**
   * Check for updates by comparing versions
   */
  private compareVersions(
    currentVersion: string,
    latestVersion: string,
  ): boolean {
    // Simple semantic version comparison (basic implementation)
    const current = currentVersion.replace(/^v/, "").split(".").map(Number);
    const latest = latestVersion.replace(/^v/, "").split(".").map(Number);

    for (let i = 0; i < Math.max(current.length, latest.length); i++) {
      const currentPart = current[i] || 0;
      const latestPart = latest[i] || 0;

      if (latestPart > currentPart) {
        return true; // Update available
      } else if (latestPart < currentPart) {
        return false; // Current is newer
      }
    }

    return false; // Versions are equal
  }

  /**
   * Check for updates manually
   */
  public async checkForUpdates(): Promise<UpdateCheckResult> {
    try {
      this.status.checking = true;
      this.status.error = undefined;

      this.notifyCallback("app-update:checking");

      const release = await this.fetchLatestRelease();

      if (release) {
        const updateAvailable = this.compareVersions(
          this.status.currentVersion,
          release.tag_name,
        );

        if (updateAvailable) {
          this.status.updateAvailable = true;
          this.status.latestVersion = release.tag_name;

          // Filter assets for web downloads (prefer .zip, .tar.gz, or installers)
          const downloadAsset = release.assets.find(
            (asset) =>
              asset.name.includes(".zip") ||
              asset.name.includes(".tar.gz") ||
              asset.name.includes("installer") ||
              asset.name.includes("setup"),
          );

          this.notifyCallback("app-update:available", {
            version: release.tag_name,
            releaseNotes: release.body,
            publishedAt: release.published_at,
            downloadUrl:
              downloadAsset?.browser_download_url || release.html_url,
            prerelease: release.prerelease,
          });
        } else {
          this.status.updateAvailable = false;
          this.notifyCallback("app-update:not-available");
        }
      }

      this.status.checking = false;

      return {
        updateAvailable: this.status.updateAvailable,
        currentVersion: this.status.currentVersion,
        latestVersion: this.status.latestVersion,
        releaseNotes: release?.body,
        downloadUrl: release?.html_url,
      };
    } catch (error) {
      this.status.checking = false;
      this.status.error = (error as Error).message;

      errorLogger.logError(error as Error, {
        service: "AutoUpdater",
        operation: "checkForUpdates",
        currentVersion: this.status.currentVersion,
      });

      this.notifyCallback("app-update:error", error);

      return {
        updateAvailable: false,
        currentVersion: this.status.currentVersion,
        error: (error as Error).message,
      };
    }
  }

  /**
   * Download the update (redirect to download URL)
   */
  public async downloadUpdate(): Promise<UpdateDownloadResult> {
    try {
      this.status.downloading = true;
      this.status.error = undefined;

      // For web apps, we redirect to the download URL rather than downloading automatically
      // The UI should prompt the user to download manually
      this.notifyCallback("app-update:download-ready", {
        downloadUrl: `https://github.com/${this.config.githubRepo}/releases/latest`,
      });

      this.status.downloading = false;
      this.status.updateDownloaded = true;

      return { success: true };
    } catch (error) {
      this.status.downloading = false;
      this.status.error = (error as Error).message;

      return { success: false, error: (error as Error).message };
    }
  }

  /**
   * For web apps, this redirects to the installation instructions
   */
  public quitAndInstall(): void {
    // In a web app, "installation" means refreshing or navigating to the new version
    window.location.reload();
  }

  /**
   * Subscribe to download progress updates
   */
  public onDownloadProgress(callback: (percent: number) => void): void {
    this.downloadProgressCallbacks.push(callback);
  }

  /**
   * Get current update status
   */
  public getStatus(): UpdateStatus {
    return { ...this.status };
  }

  /**
   * Initialize the updater (check for updates on startup if configured)
   */
  public async initialize(): Promise<void> {
    if (this.config.checkOnStartup) {
      try {
        await this.checkForUpdates();
      } catch (error) {
        errorLogger.logError(error as Error, {
          service: "AutoUpdater",
          operation: "initialize",
        });
      }
    }
  }

  /**
   * Set a callback for update notifications
   */
  public setNotificationCallback(callback: UpdateNotificationCallback): void {
    this.notificationCallback = callback;
  }

  /**
   * Get the update source being used
   */
  public getUpdateSource(): string {
    if (this.config.githubRepo) {
      return `GitHub: ${this.config.githubRepo}`;
    }
    return "github";
  }

  /**
   * Check if auto-updates are enabled
   */
  public isEnabled(): boolean {
    // Web apps can always check for updates, but installation is manual
    return true;
  }

  /**
   * Clean up resources
   */
  public dispose(): void {
    if (this.updateCheckTimer) {
      clearInterval(this.updateCheckTimer);
      this.updateCheckTimer = null;
    }
  }
}

File: backend/services/auto_updater.py
"""
AutoUpdater service for checking and downloading application updates from GitHub.

Migrated from: src/services/AutoUpdater.ts

Features:
- Check for updates from GitHub releases
- Download updates with progress tracking
- Version comparison using semantic versioning
- Channel support (stable, beta, alpha)
- Configurable update server URL
- Progress callbacks for download tracking
- Comprehensive error handling and logging
- Async operations for non-blocking updates

Usage:
    from backend.services.auto_updater import AutoUpdater, AutoUpdaterConfig

    config = AutoUpdaterConfig(
        check_on_startup=True,
        channel="stable"
    )

    updater = AutoUpdater(
        repo="owner/repo",
        current_version="1.0.0",
        config=config
    )

    # Check for updates
    result = await updater.check_for_updates()
    if result.update_available:
        print(f"Update available: {result.latest_version}")

        # Download update
        download_result = await updater.download_update(result.update_info)
        if download_result.success:
            print("Update downloaded successfully")

Security:
- HTTPS-only connections
- SHA256 checksum verification for downloads
- Version validation to prevent downgrade attacks
- Signed release verification (if GitHub signing is enabled)
"""

import asyncio
import hashlib
import json
import logging
import os
import re
from dataclasses import dataclass, field, asdict
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from typing import Optional, List, Dict, Any, Callable, Tuple
from urllib.parse import urlparse

import httpx
from packaging.version import Version, InvalidVersion

# Configure logger
logger = logging.getLogger(__name__)


# ============================================================================
# TYPE DEFINITIONS & DATA CLASSES
# ============================================================================


class UpdateChannel(str, Enum):
    """Update channel enumeration."""
    STABLE = "stable"
    BETA = "beta"
    ALPHA = "alpha"


@dataclass
class AutoUpdaterConfig:
    """Configuration for auto-updater."""
    check_on_startup: bool = True
    update_server_url: Optional[str] = None
    channel: UpdateChannel = UpdateChannel.STABLE
    github_token: Optional[str] = None  # For private repositories
    timeout_seconds: int = 30
    max_retries: int = 3


@dataclass
class UpdateInfo:
    """Information about an available update."""
    version: str
    download_url: str
    release_notes: str
    published_at: str
    size_bytes: Optional[int] = None
    checksum_sha256: Optional[str] = None
    is_prerelease: bool = False
    tag_name: str = ""

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return asdict(self)


@dataclass
class ProgressInfo:
    """Download progress information."""
    transferred: int
    total: int
    percent: float
    bytes_per_second: float

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return asdict(self)


@dataclass
class UpdateStatus:
    """Current status of the updater."""
    current_version: str
    latest_version: Optional[str] = None
    checking: bool = False
    update_available: bool = False
    downloading: bool = False
    update_downloaded: bool = False
    download_progress: Optional[float] = None
    error: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return asdict(self)


@dataclass
class UpdateCheckResult:
    """Result of checking for updates."""
    update_available: bool
    current_version: str
    latest_version: Optional[str] = None
    update_info: Optional[UpdateInfo] = None
    error: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        result = {
            "update_available": self.update_available,
            "current_version": self.current_version,
            "latest_version": self.latest_version,
            "error": self.error
        }
        if self.update_info:
            result["update_info"] = self.update_info.to_dict()
        return result


@dataclass
class UpdateDownloadResult:
    """Result of downloading an update."""
    success: bool
    file_path: Optional[str] = None
    error: Optional[str] = None
    checksum_verified: bool = False

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return asdict(self)


# ============================================================================
# AUTO UPDATER SERVICE
# ============================================================================


class AutoUpdaterError(Exception):
    """Auto updater error exception."""
    pass


class AutoUpdater:
    """
    Auto updater service for checking and downloading application updates.

    Supports GitHub releases with semantic versioning, progress tracking,
    and configurable update channels (stable, beta, alpha).
    """

    # Constants
    GITHUB_API_BASE = "https://api.github.com"
    USER_AGENT = "JusticeCompanion-AutoUpdater/1.0"

    def __init__(
        self,
        repo: str,
        current_version: str,
        config: Optional[AutoUpdaterConfig] = None,
        audit_logger: Optional[Any] = None
    ):
        """
        Initialize auto updater.

        Args:
            repo: GitHub repository in format "owner/repo"
            current_version: Current application version (semver format)
            config: Auto updater configuration
            audit_logger: Optional audit logger instance
        """
        self.repo = repo
        self.current_version = current_version
        self.config = config or AutoUpdaterConfig()
        self.audit_logger = audit_logger

        # Validate repository format
        if not re.match(r'^[\w-]+/[\w-]+$', repo):
            raise AutoUpdaterError(f"Invalid repository format: {repo}. Expected 'owner/repo'")

        # Validate current version
        try:
            Version(current_version)
        except InvalidVersion:
            raise AutoUpdaterError(f"Invalid version format: {current_version}")

        # Initialize status
        self.status = UpdateStatus(current_version=current_version)

        # Download progress callbacks
        self.download_progress_callbacks: List[Callable[[float], None]] = []

        # HTTP client
        self.client: Optional[httpx.AsyncClient] = None

        logger.info(
            f"[AutoUpdater] Initialized for {repo} (current: {current_version}, "
            f"channel: {self.config.channel.value})"
        )

    async def __aenter__(self):
        """Async context manager entry."""
        await self._ensure_client()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit."""
        await self.close()

    async def _ensure_client(self):
        """Ensure HTTP client is initialized."""
        if not self.client:
            headers = {
                "User-Agent": self.USER_AGENT,
                "Accept": "application/vnd.github.v3+json"
            }

            # Add GitHub token if provided (for private repos)
            if self.config.github_token:
                headers["Authorization"] = f"token {self.config.github_token}"

            self.client = httpx.AsyncClient(
                headers=headers,
                timeout=self.config.timeout_seconds,
                follow_redirects=True
            )

    async def close(self):
        """Close HTTP client."""
        if self.client:
            await self.client.aclose()
            self.client = None

    def compare_versions(self, v1: str, v2: str) -> int:
        """
        Compare two semantic version strings.

        Args:
            v1: First version string
            v2: Second version string

        Returns:
            -1 if v1 < v2
             0 if v1 == v2
             1 if v1 > v2

        Raises:
            AutoUpdaterError: If version strings are invalid
        """
        try:
            version1 = Version(v1)
            version2 = Version(v2)

            if version1 < version2:
                return -1
            elif version1 > version2:
                return 1
            else:
                return 0
        except InvalidVersion as e:
            raise AutoUpdaterError(f"Invalid version format: {e}")

    def _is_version_compatible(self, version_str: str) -> bool:
        """
        Check if version is compatible with configured channel.

        Args:
            version_str: Version string to check

        Returns:
            True if version is compatible with channel
        """
        try:
            version = Version(version_str)

            # Stable channel: only stable releases (no pre-release)
            if self.config.channel == UpdateChannel.STABLE:
                return not version.is_prerelease

            # Beta channel: stable + beta releases
            elif self.config.channel == UpdateChannel.BETA:
                if version.is_prerelease:
                    return any(pre[0] in ('beta', 'b') for pre in version.pre or [])
                return True

            # Alpha channel: all releases
            elif self.config.channel == UpdateChannel.ALPHA:
                return True

            return False
        except InvalidVersion:
            return False

    async def _fetch_github_releases(self) -> List[Dict[str, Any]]:
        """
        Fetch releases from GitHub API.

        Returns:
            List of release objects from GitHub API

        Raises:
            AutoUpdaterError: If API request fails
        """
        await self._ensure_client()

        url = f"{self.GITHUB_API_BASE}/repos/{self.repo}/releases"

        for attempt in range(self.config.max_retries):
            try:
                response = await self.client.get(url)
                response.raise_for_status()

                releases = response.json()
                logger.info(f"[AutoUpdater] Fetched {len(releases)} releases from GitHub")

                return releases

            except httpx.HTTPStatusError as e:
                if e.response.status_code == 404:
                    raise AutoUpdaterError(f"Repository not found: {self.repo}")
                elif e.response.status_code == 403:
                    raise AutoUpdaterError("GitHub API rate limit exceeded")
                else:
                    logger.warning(
                        f"[AutoUpdater] HTTP error fetching releases (attempt {attempt + 1}): {e}"
                    )
                    if attempt == self.config.max_retries - 1:
                        raise AutoUpdaterError(f"Failed to fetch releases: {e}")

            except httpx.RequestError as e:
                logger.warning(
                    f"[AutoUpdater] Network error fetching releases (attempt {attempt + 1}): {e}"
                )
                if attempt == self.config.max_retries - 1:
                    raise AutoUpdaterError(f"Network error: {e}")

            # Exponential backoff
            if attempt < self.config.max_retries - 1:
                await asyncio.sleep(2 ** attempt)

        return []

    def _parse_release_to_update_info(self, release: Dict[str, Any]) -> Optional[UpdateInfo]:
        """
        Parse GitHub release object to UpdateInfo.

        Args:
            release: GitHub release object

        Returns:
            UpdateInfo object or None if no suitable asset found
        """
        # Extract version from tag name (remove 'v' prefix if present)
        tag_name = release.get("tag_name", "")
        version_str = tag_name.lstrip("v")

        # Validate version
        if not self._is_version_compatible(version_str):
            return None

        # Find suitable asset (platform-specific)
        # For now, just get the first asset
        assets = release.get("assets", [])
        if not assets:
            logger.warning(f"[AutoUpdater] Release {version_str} has no assets")
            return None

        # TODO: Filter by platform (Windows .exe, macOS .dmg, Linux .AppImage)
        asset = assets[0]

        return UpdateInfo(
            version=version_str,
            download_url=asset.get("browser_download_url", ""),
            release_notes=release.get("body", ""),
            published_at=release.get("published_at", ""),
            size_bytes=asset.get("size"),
            tag_name=tag_name,
            is_prerelease=release.get("prerelease", False)
        )

    async def check_for_updates(self) -> UpdateCheckResult:
        """
        Check for available updates.

        Returns:
            UpdateCheckResult with update information
        """
        try:
            self.status.checking = True
            self.status.error = None

            logger.info(f"[AutoUpdater] Checking for updates (current: {self.current_version})")

            # Fetch releases from GitHub
            releases = await self._fetch_github_releases()

            # Find latest compatible version
            latest_update: Optional[UpdateInfo] = None

            for release in releases:
                update_info = self._parse_release_to_update_info(release)
                if not update_info:
                    continue

                # Check if this version is newer than current
                if self.compare_versions(update_info.version, self.current_version) > 0:
                    # Check if this is newer than previously found update
                    if not latest_update or self.compare_versions(
                        update_info.version, latest_update.version
                    ) > 0:
                        latest_update = update_info

            # Update status
            self.status.checking = False

            if latest_update:
                self.status.update_available = True
                self.status.latest_version = latest_update.version

                logger.info(
                    f"[AutoUpdater] Update available: {latest_update.version} "
                    f"(current: {self.current_version})"
                )

                # Log to audit
                if self.audit_logger:
                    await self.audit_logger.log_event(
                        action="update_check",
                        details={
                            "current_version": self.current_version,
                            "latest_version": latest_update.version,
                            "update_available": True
                        }
                    )

                return UpdateCheckResult(
                    update_available=True,
                    current_version=self.current_version,
                    latest_version=latest_update.version,
                    update_info=latest_update
                )
            else:
                self.status.update_available = False

                logger.info(f"[AutoUpdater] No updates available (current: {self.current_version})")

                return UpdateCheckResult(
                    update_available=False,
                    current_version=self.current_version
                )

        except AutoUpdaterError as e:
            self.status.checking = False
            self.status.error = str(e)

            logger.error(f"[AutoUpdater] Error checking for updates: {e}")

            return UpdateCheckResult(
                update_available=False,
                current_version=self.current_version,
                error=str(e)
            )

        except Exception as e:
            self.status.checking = False
            self.status.error = str(e)

            logger.exception(f"[AutoUpdater] Unexpected error checking for updates: {e}")

            return UpdateCheckResult(
                update_available=False,
                current_version=self.current_version,
                error=f"Unexpected error: {e}"
            )

    async def download_update(
        self,
        update_info: UpdateInfo,
        output_path: Optional[str] = None
    ) -> UpdateDownloadResult:
        """
        Download update file.

        Args:
            update_info: Update information
            output_path: Optional output file path (default: temp directory)

        Returns:
            UpdateDownloadResult with download status
        """
        try:
            self.status.downloading = True
            self.status.error = None
            self.status.download_progress = 0.0

            logger.info(f"[AutoUpdater] Downloading update {update_info.version}")

            # Determine output path
            if not output_path:
                # Extract filename from URL
                parsed_url = urlparse(update_info.download_url)
                filename = os.path.basename(parsed_url.path)

                # Use temp directory
                temp_dir = Path.home() / ".cache" / "justice-companion" / "updates"
                temp_dir.mkdir(parents=True, exist_ok=True)

                output_path = str(temp_dir / filename)

            # Ensure output directory exists
            Path(output_path).parent.mkdir(parents=True, exist_ok=True)

            await self._ensure_client()

            # Download with progress tracking
            start_time = datetime.now(timezone.utc)
            bytes_downloaded = 0

            async with self.client.stream("GET", update_info.download_url) as response:
                response.raise_for_status()

                total_size = int(response.headers.get("content-length", 0))

                # Verify size matches if provided
                if update_info.size_bytes and total_size != update_info.size_bytes:
                    raise AutoUpdaterError(
                        f"Size mismatch: expected {update_info.size_bytes}, got {total_size}"
                    )

                # Download and write to file
                hasher = hashlib.sha256()

                with open(output_path, "wb") as f:
                    async for chunk in response.aiter_bytes(chunk_size=8192):
                        f.write(chunk)
                        hasher.update(chunk)

                        bytes_downloaded += len(chunk)

                        # Calculate progress
                        if total_size > 0:
                            percent = (bytes_downloaded / total_size) * 100
                            self.status.download_progress = percent

                            # Calculate speed
                            elapsed = (datetime.now(timezone.utc) - start_time).total_seconds()
                            bytes_per_second = bytes_downloaded / elapsed if elapsed > 0 else 0

                            # Notify callbacks
                            progress_info = ProgressInfo(
                                transferred=bytes_downloaded,
                                total=total_size,
                                percent=percent,
                                bytes_per_second=bytes_per_second
                            )

                            for callback in self.download_progress_callbacks:
                                try:
                                    callback(percent)
                                except Exception as e:
                                    logger.warning(
                                        f"[AutoUpdater] Error in progress callback: {e}"
                                    )

            # Verify checksum if provided
            checksum_verified = False
            if update_info.checksum_sha256:
                calculated_checksum = hasher.hexdigest()
                if calculated_checksum != update_info.checksum_sha256:
                    # Delete corrupted file
                    os.remove(output_path)
                    raise AutoUpdaterError(
                        f"Checksum mismatch: expected {update_info.checksum_sha256}, "
                        f"got {calculated_checksum}"
                    )
                checksum_verified = True
                logger.info("[AutoUpdater] Checksum verified successfully")

            self.status.downloading = False
            self.status.update_downloaded = True
            self.status.download_progress = 100.0

            logger.info(f"[AutoUpdater] Update downloaded successfully to {output_path}")

            # Log to audit
            if self.audit_logger:
                await self.audit_logger.log_event(
                    action="update_download",
                    details={
                        "version": update_info.version,
                        "size_bytes": bytes_downloaded,
                        "checksum_verified": checksum_verified
                    }
                )

            return UpdateDownloadResult(
                success=True,
                file_path=output_path,
                checksum_verified=checksum_verified
            )

        except AutoUpdaterError as e:
            self.status.downloading = False
            self.status.error = str(e)

            logger.error(f"[AutoUpdater] Error downloading update: {e}")

            return UpdateDownloadResult(
                success=False,
                error=str(e)
            )

        except Exception as e:
            self.status.downloading = False
            self.status.error = str(e)

            logger.exception(f"[AutoUpdater] Unexpected error downloading update: {e}")

            return UpdateDownloadResult(
                success=False,
                error=f"Unexpected error: {e}"
            )

    def on_download_progress(self, callback: Callable[[float], None]):
        """
        Subscribe to download progress updates.

        Args:
            callback: Function to call with progress percentage (0-100)
        """
        self.download_progress_callbacks.append(callback)

    def get_status(self) -> UpdateStatus:
        """
        Get current update status.

        Returns:
            Copy of current UpdateStatus
        """
        return UpdateStatus(**asdict(self.status))

    async def initialize(self):
        """
        Initialize updater and optionally check for updates on startup.
        """
        if self.config.check_on_startup:
            try:
                logger.info("[AutoUpdater] Checking for updates on startup")
                await self.check_for_updates()
            except Exception as e:
                logger.error(f"[AutoUpdater] Error during startup update check: {e}")

    def get_update_source(self) -> str:
        """
        Get the update source being used.

        Returns:
            Update source ("github" or custom URL)
        """
        if self.config.update_server_url:
            return self.config.update_server_url
        return f"github:{self.repo}"

    def is_enabled(self) -> bool:
        """
        Check if auto-updates are enabled.

        Returns:
            True if updates are enabled (production environment)
        """
        # In Python backend, we don't have NODE_ENV
        # Check if we're in debug mode or have updates explicitly disabled
        return not os.getenv("DISABLE_AUTO_UPDATES", "").lower() in ("true", "1", "yes")


# ============================================================================
# HELPER FUNCTIONS
# ============================================================================


def format_file_size(size_bytes: int) -> str:
    """
    Format file size in human-readable format.

    Args:
        size_bytes: Size in bytes

    Returns:
        Formatted string (e.g., "10.5 MB")
    """
    for unit in ["B", "KB", "MB", "GB"]:
        if size_bytes < 1024.0:
            return f"{size_bytes:.1f} {unit}"
        size_bytes /= 1024.0
    return f"{size_bytes:.1f} TB"


def validate_version(version_str: str) -> bool:
    """
    Validate semantic version string.

    Args:
        version_str: Version string to validate

    Returns:
        True if valid semantic version
    """
    try:
        Version(version_str)
        return True
    except InvalidVersion:
        return False
